test/drop_test/ffi.py

```diff
@@ -173,6 +173,40 @@ def secret(self):
         return self.this.privkey
 
 
+class FdResolver:
+    def __init__(self):
+        self._files = []
+        self._cached = {}
+
+    def callback(self, ctx, uri):
+        uri: str = uri.decode("utf-8")
+        path = uri.removeprefix("content://")
```

Comment:
`content://` or `cached` or `new` are entirely this test suite's things. I suggest simplifying the state machine and just speaking using file paths as those are pretty unique and enough for test cases. This way this callback would be very simple by just opening the path and returning a file descriptor.

Reply:
Some test cases resue the FDs so the caching mechanism must be present

Reply:
the caching mechanism is fine but it can be as simple as just a hashmap with path as a key and fd as a value and if it's there - it's cached the way I see it

Reply:
Yes, it is there. Some tests are refusing FDs and some do not. So I've devised these `cached/` and `new/` to instruct the callback to do the right thing

Reply:
Oh I see, if it's a callback then it needs to differentiate between cached and non-cached paths. Makes sense